src/lib/Libifl/pbsD_deljoblist.c

```diff
@@ -46,160 +46,554 @@
+ * @retval list of error objects from server (or NULL)
+ */
+static struct batch_deljob_status *
+deljob_to_server(int fd, int function, char **jobids, int numids, char *extend)
+{
+	if (send_deljob(fd, function, jobids, numids, extend))
 		return NULL;
 
-	return rbsp;
+	return recv_deljob(fd);
 }
 
 /**
- * @brief
- *      -encode a Delete Job List Batch Request
+ * @brief	Broadcast deljob request to other servers in a multi-server set up
+ * 			This handles the cases where a job was moved to another server
+ *
+ * @param[in] c - connection handle (virtual)
+ * @param[in] function - req type
+ * @param[in] jobids - the job id list
+ * @param[in] numjids - the count of job ids
+ * @param[in] extend - extend string for req
  *
- * @par Functionality:
- *              This request is used for delete operation.
- *              
+ * @return struct batch_deljob_status *
+ * @retval list of jobs not found on any server
+ */
+static struct batch_deljob_status *
+broadcast_deljob(int c, int function, char **jobids, int numids, char *extend)
+{
+	int *skip_list;
+	svr_conn_t **conns = get_conn_svr_instances(c);
+	int num_svrs = get_num_servers();
+	int i;
+	struct batch_deljob_status **svr_retlist = NULL;
+	struct batch_deljob_status *retlist = NULL;
+	struct batch_deljob_status *iter_list = NULL;
+	struct batch_deljob_status *prev = NULL;
+	struct batch_deljob_status *next = NULL;
+
+	if (conns == NULL) {
+		pbs_errno = PBSE_NOCONNECTS;
+		return NULL;
+	}
+
+	skip_list = calloc(num_svrs, sizeof(int));
+	if (skip_list == NULL) {
+		pbs_errno = PBSE_SYSTEM;
+		return NULL;
+	}
+
+	svr_retlist = calloc(num_svrs, sizeof(struct batch_deljob_status *));
+	if (svr_retlist == NULL) {
+		free(skip_list);
+		pbs_errno = PBSE_SYSTEM;
+		return NULL;
+	}
+
+	/* Broadcast the jids to all servers */
+	for (i = 0; i < num_svrs; i++) {
+		if (send_deljob(conns[i]->sd, function, jobids, numids, extend))
+			skip_list[i] = 1;
+	}
+	for (i = 0; i < num_svrs; i++) {
+		if (!skip_list[i]) {
+			struct batch_deljob_status *ret = NULL;
+
+			ret = recv_deljob(conns[i]->sd);
+			svr_retlist[i] = ret;
+		}
+	}
+
+	/* The list of ids we return has to be the intersection of the lists from all servers
+	 * Use the first server's list as the intersection to be returned
+	 */
+	for (iter_list = svr_retlist[0]; iter_list != NULL; prev = iter_list, iter_list = next) {
+		int job_count = 1;
+		struct batch_deljob_status *iter_list2;
+		next = iter_list->next;
+
+		/* Make sure that this jid is in ALL of the other lists as well */
+		for (i = 1; i < num_svrs; i++) {
+			for (iter_list2 = svr_retlist[i]; iter_list2 != NULL; iter_list2 = iter_list2->next) {
+				if (strcmp(iter_list2->name, iter_list->name) == 0) {
+					job_count++;
+					break;
+				}
+			}
+		}
+
+		if (job_count != num_svrs) {	/* Job was found (and deleted) on at least 1 server */
+			/* Remove the job from the list to return */
+			if (prev != NULL)
+				prev->next = next;
+			else
+				svr_retlist[0] = next;
```

Comment:
Nithin has a good point. Do we really have to report users about jobs that are not present in any of the servers? Can we just not iterate through the lists and report jobs that servers fail to delete because of other reasons? Those are the actual jobs a user would care about, isn't it?

Reply:
I agree, we should care more about legit failures to delete the job. But here, we are removing the jobids which were successfully deleted on other servers, so it's perfectly safe to ignore their error message. qdel is returned the complete list of all jids with error codes if they couldn't be deleted on any servers.

Reply:
on second thought, to return the accurate error code, I'll need to add a check inside the loop above this, that's where we go through the lists of all servers, so thanks for pointing this out.

Reply:
Here is what I understand from what I read, please correct me if I'm wrong - 
If you have 'm' servers and they reply with 'n' number of jobs that couldn't be deleted, you are looping through all servers and the jobs they reported, so the complexity is close to m*n^2.
Can't you just loop through all the lists one after another and just gather jobs that have code other than unknown job id? that will result in lowering down complexity to m*n, isn't it? I think it is safe to do as long as you don't want to return back jobs with unknown job error code.

Reply:
I agree with the approach. Just noting that we also might need to report back unknown job id, if that is what you heard back from all the servers.

Reply:
that's the thing, we do need to report back unknowns as well if no server found the job. The only way to know if a job was found on others servers or not is by scanning each server's list, which adds that extra n and makes it mn^2. @nithinj did you have an algorithm in mind which reduces the time complexity?

In practice, the broadcast list should be small, we anticipate a small fraction of jobs to be moved between servers, so I don't think we have to worry about this, but if there's an obviously more performant algorithm, please do let me know.

Reply:
You could use avl indexing to reduce complexity. But that seems unnecessary if the number of jobs is going to be a small percentage. Also, deljoblist is batching jobs before sending them to the server. 
One optimization you can do in your current approach is to break the inner loop of iterating over nsvrs when you encounter an rc != UNKWN_ID or when you don't find the job in a servers list.

Reply:
"You could use avl indexing to reduce complexity" I agree with you that it's not worth it, but I did think about it and didn't really see how it would help, so out of curiosity, can you please tell me what you had in mind?

"One optimization you can do in your current approach is to break the inner loop of iterating over nsvrs when you encounter an rc != UNKWN_ID or when you don't find the job in a servers list."
ok ya, I should definitely do that, thanks for the suggestion

Reply:
You need to walk over all the jobs returned by servers anyway. That is mn.
When we encounter a new jobid, add it to the index.
If the index contains the ID, update it with rc and the number of times it had been repeated.
Before returning the final list remove every element which has not repeated m times and having rc as unknwn_ID.
Since we are accessing from index for every m*n, the complexity will be mnlogn

Reply:
do we have to report unknown job ids? qdel in its current form does not report that.

Reply:
oh, that's interesting. If we don't need to then we can simplify this. Thanks!

Reply:
Your process_deljobstat function has this check. If deljoblist returns PBSE_UNKJOBID, which means the job got moved to a different cluster. It sends a locatejob to identify where the job has been moved to and sends a delete request to that cluster.

Reply:
Yeah, The current qdel client would need to know the return code PBSE_UNKJOBID for the moved job, so that qdel would locate it and send the delete request.

Reply:
oh right, ill need to revert this change then, thanks for catching that guys